src/Platform/Microsoft.Testing.Platform/UI/ConsoleLogger.cs

```diff
@@ -0,0 +1,701 @@
+
+        string atResourceName = "Word_At";
+        string inResourceName = "StackTrace_InFileLineNumber";
+
+        string? atString = null;
+        string? inString = null;
+
+        // Grab words from localized resource, in case the stack trace is localized.
+        try
+        {
+            // Get these resources: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/Resources/Strings.resx
+#pragma warning disable RS0030 // Do not use banned APIs
+            MethodInfo? getResourceStringMethod = typeof(Environment).GetMethod("GetResourceString", BindingFlags.Static | BindingFlags.NonPublic, null, [typeof(string)], null);
+#pragma warning restore RS0030 // Do not use banned APIs
+            if (getResourceStringMethod is not null)
+            {
+                // <value>at</value>
+                atString = (string?)getResourceStringMethod.Invoke(null, [atResourceName]);
+
+                // <value>in {0}:line {1}</value>
+                inString = (string?)getResourceStringMethod.Invoke(null, [inResourceName]);
+            }
+        }
+        catch
+        {
+            // If we fail, populate the defaults below.
+        }
+
+        atString = atString == null || atString == atResourceName ? "at" : atString;
+        inString = inString == null || inString == inResourceName ? "in {0}:line {1}" : inString;
+
+        string inPattern = string.Format(CultureInfo.InvariantCulture, inString, "(?<file>.+)", @"(?<line>\d+)");
+
+        s_regex = new Regex(@$"^   {atString} (?<code>.+) {inPattern}$", RegexOptions.Compiled | RegexOptions.ExplicitCapture, matchTimeout: TimeSpan.FromSeconds(1));
+        return s_regex;
+    }
+#endif
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ConsoleLogger"/> class with custom terminal and manual refresh for testing.
+    /// </summary>
+    internal ConsoleLogger(IConsole console, ConsoleLoggerOptions options)
+    {
+        _options = options;
+
+        bool showProgress = _options.ShowProgress;
+        ConsoleWithProgress consoleWithProgress;
+        if (!_options.UseAnsi)
+        {
+            consoleWithProgress = new ConsoleWithProgress(new AnsiTerminal(console, _options.BaseDirectory), showProgress, updateEvery: 33);
+        }
+        else
+        {
+            // Autodetect.
+            (bool consoleAcceptsAnsiCodes, bool _, uint? originalConsoleMode) = NativeMethods.QueryIsScreenAndTryEnableAnsiColorCodes();
+            _originalConsoleMode = originalConsoleMode;
+            consoleWithProgress = consoleAcceptsAnsiCodes
+                ? new ConsoleWithProgress(new AnsiTerminal(console, _options.BaseDirectory), showProgress, updateEvery: 33)
+                : new ConsoleWithProgress(new NonAnsiTerminal(console), showProgress, updateEvery: 3_000);
+        }
+
+        _consoleWithProgress = consoleWithProgress;
+    }
+
+    public void TestExecutionStarted(DateTimeOffset testStartTime, int workerCount)
```

Comment:
It's called from only one place with `1` for worker, do we need the parameter? What does it represent?

Reply:
This will be used by dotnet test when we have multiple projects running via this same logger, when there is the new protocol between dotnet test and test `.exe`.

Reply:
Is it responsibility of the caller or this class to know how many state object should handle the progress?

Reply:
The caller tells us what is the level of parallelism, so we can know how many projects will run at the same time at max, and make UI choices based on that. e.g. we could write: "Starting run in parallel on 10 workers".

Actually managing the states is the responsibility of this console with progress.